### ISSUE TITLE
Xamarin TestCloud task wildcard matching fails as release step

### ISSUE THREAD

**Issue by ColinDabritz** [Microsoft/vsts-tasks#2067]
HAS WORKAROUND - Use powershell to fully resolve paths in a separate step, see my post below for details.

Symptoms:
The Xamarin TestCloud step works correct as a build step, but as a similarly configured release step appears to not handle wildcards in the App File and test-cloud.exe Location paths.

For example, I have a sample release step with the following App File path:
```
$(System.DefaultWorkingDirectory)/**/*.ipa
```

gets the following error while running (against a hosted build agent):
```
2016-07-12T19:22:15.4082748Z ##[error]No matching app files were found with search pattern: C:\a\958d17f14\**\*.ipa
 
2016-07-12T19:22:15.4092746Z ##[debug]task result: Failed
```

We definitely have the file in place, as shown by the output from this inline PowerShell script step:
```
echo "Build Directory (Probably blank in release steps): "
echo $env:AGENT_BUILDDIRECTORY
echo "----------"
$ipa = Resolve-Path SomeProject.iOS\bin\iPhone\UITest\**\*.ipa
echo "Full IPA path:"
echo $ipa.Path
```

Which gives output:
``` 
2016-07-12T19:22:15.0792750Z Build Directory (Probably blank in release steps): 
 
2016-07-12T19:22:15.0802747Z ----------
 
2016-07-12T19:22:15.0822738Z Full IPA path:
 
2016-07-12T19:22:15.0842750Z C:\a\958d17f14\Issue Repro Build\drop\SomeProject.iOS\bin\iPhone\UITest\SomeProjectiOS 2016-06-30 09-39-16\SomeProject.ipa
```

I would expect release tasks to behave similarly to build tasks.

Some helpful pointers from support showed that this repo was online, and we tracked the error down to here:

https://github.com/Microsoft/vsts-tasks/blob/master/Tasks/XamarinTestCloud/xamarintestcloud.ts#L43

```
else {
    // Find app files matching the specified pattern    
    tl.debug('Pattern found in app parameter');
    var buildFolder = tl.getVariable('agent.buildDirectory');
    var allappFiles = tl.find(buildFolder);
    var appFiles = tl.match(allappFiles, app, {matchBase: true});

    // Fail if no matching app files were found
    if (!appFiles || appFiles.length == 0) {
        onError('No matching app files were found with search pattern: ' + app);
    }
```

In this code it appears that the wild-card matching branch is assuming we are running in a build scenario, as it explicitly gets the files only from the agent.BuildDirectory variable. This likely produces no results in a release scenario.

Here are some approaches we came up with for addressing this issue:
* We could detect release vs build variables, and use whichever has a value
* We could instead base the search on the working directory
* We could split the path before the first wildcard and search below there for matching purposes

Definitely some pros and cons to these.

We are intended to follow up with a fix, which we are creating as a live hands-on coding session at the DotNetFringe conference! fun times. Expect a pull request for this issue later today, if we can figure out how to fix and test this problem. otherwise we will post a follow up with how far we get.

**Comment by ColinDabritz** [comment]
Ok, it sounds like there may be bigger questions to resolve around tooling and related challenges. To get more information about exactly what was available I put together some variable options and ran an info script in a build and a release.

I collected these options to test from:

https://www.visualstudio.com/docs/build/define/variables
https://www.visualstudio.com/docs/release/author-release-definition/understanding-tasks#predefvariables

and reading the source of exiting tasks in this repo.

Here is the script:
```
echo ("System.DefaultWorkingDirectory: " + $env:SYSTEM_DEFAULTWORKINGDIRECTORY)
echo ("agent.buildDirectory: " + $env:AGENT_BUILDDIRECTORY)
echo ("build.sourceDirectory: " + $env:BUILD_SOURCEDIRECTORY)
echo ("Agent.WorkingDirectory: " + $env:AGENT_WORKINGDIRECTORY)
echo ("Agent.RootDirectory: " + $env:AGENT_ROOTDIRECTORY)
```

and the results from a build:
```
2016-07-13T03:38:37.0087969Z . 'C:\Users\buildguest\AppData\Local\Temp\5b193b00-de51-425d-bf4b-77dd0697d383.ps1' 
2016-07-13T03:38:37.0367959Z C:\Windows\System32\WindowsPowerShell\v1.0\powershell.exe -NoLogo -NoProfile -NonInteractive -ExecutionPolicy Unrestricted -Command "$env:PSModulePath= $env:PSModulePath + ';' + $([system.io.path]::combine($env:AGENT_HOMEDIRECTORY, 'agent\worker\modules')); try { [System.Security.Cryptography.ProtectedData] | Out-Null } catch { Write-Verbose 'Adding assemly: System.Security' ; Add-Type -AssemblyName 'System.Security' ; [System.Security.Cryptography.ProtectedData] | Out-Null } ; Invoke-Expression -Command ([System.Text.Encoding]::UTF8.GetString([System.Security.Cryptography.ProtectedData]::Unprotect([System.Convert]::FromBase64String('AQAAANCMnd8BFdERjHoAwE/Cl+sBAAAAg1uEV+Rrg0SAuwVR31VyzgAAAAACAAAAAAAQZgAAAAEAACAAAAB44TCLN2CiRaOpJGHVRmV41rDiuB71oTpmv/A8qUy1igAAAAAOgAAAAAIAACAAAAB0aYezG0xh9Vz6M/cMrfg0Q4K6EUKIg0d6HejzHdMv52AAAACyLQPAh7rR+TZYbAIHZBtN2U1g396+HSdFG+a1CyB/EYSO9C6tRSUjefEburq8nzmitudZNyohva812eFtFhJ6S2bKCfN9fauYqKV8qoXgmF/ehcOPeeLbbvCyEAuCwxZAAAAA0+44u37hcE7dWA+rvt3MH7bd0/I5PuYOLhqH3px/DjXJINONGmESoyBu8rCYgzfY7qNwI2dKXm53Ot0CbOgC9Q=='), [System.Convert]::FromBase64String('LzRbiKsSF9pGXw9oPEaSiA=='), [System.Security.Cryptography.DataProtectionScope]::CurrentUser))) ; if (!(Test-Path -LiteralPath variable:\LastExitCode)) { Write-Verbose 'Last exit code is not set.' } else { Write-Verbose ('$LastExitCode: {0}' -f $LastExitCode) ; exit $LastExitCode }"
2016-07-13T03:38:37.0367959Z Executing the following powershell script. (workingFolder = C:\a\1\s)
2016-07-13T03:38:37.0377969Z C:\Users\buildguest\AppData\Local\Temp\5b193b00-de51-425d-bf4b-77dd0697d383.ps1 
2016-07-13T03:38:40.3338636Z System.DefaultWorkingDirectory: C:\a\1\s
2016-07-13T03:38:40.3338636Z agent.buildDirectory: C:\a\1
2016-07-13T03:38:40.3338636Z build.sourceDirectory: 
2016-07-13T03:38:40.3348637Z Agent.WorkingDirectory: C:\a\SourceRootMapping\bd6fbf29-da23-43c1-9696-de05877354e2\Job-50d9aba8-20ff-482f-85df-848a706e01f8
2016-07-13T03:38:40.3348637Z Agent.RootDirectory: C:\a
```
Where a valid file location would be something like:
`C:/a/1/a/drop/SomeProject.iOS/bin/iPhone/UITest/SomeProjectiOS 2016-06-30 09-39-16/SomeProject.ipa`

and a release:
```
2016-07-13T03:38:17.8136606Z . 'C:\Users\buildguest\AppData\Local\Temp\b30d57c4-68c1-42f9-9013-e1eb2d552b98.ps1' 
2016-07-13T03:38:17.8136606Z Executing the following powershell script. (workingFolder = C:\a\958d17f14)
2016-07-13T03:38:17.8146619Z C:\Users\buildguest\AppData\Local\Temp\b30d57c4-68c1-42f9-9013-e1eb2d552b98.ps1 
2016-07-13T03:38:18.3106622Z System.DefaultWorkingDirectory: C:\a\958d17f14
2016-07-13T03:38:18.3126623Z agent.buildDirectory: 
2016-07-13T03:38:18.3126623Z build.sourceDirectory: 
2016-07-13T03:38:18.3136630Z Agent.WorkingDirectory: C:\a\SourceRootMapping\bd6fbf29-da23-43c1-9696-de05877354e2\Job-985d770e-f605-4cb3-9f1c-74e5504a6336
2016-07-13T03:38:18.3146624Z Agent.RootDirectory: C:\a
```
where an example file would be under:
`C:\a\958d17f14\Issue Repro Build\drop\SomeProject.iOS\bin\iPhone\UITest\SomeProjectiOS 2016-06-30 09-39-16\SomeProject.ipa`

Both examples were run on hosted build agents.

These values offer some possibilities, although none of them are clearly correct to me.

The system default working directory is close, but in the build case it is the /s folder, which I believe hosts sources specifically, where as in the release cases it appears to be the 'root' folder for the release.

Changing to the working directory could break builds that rely on files below the agent folder that get generated during the process, but do not reside in the source, for example, the artifacts directory.

the build and source directories are empty in the release case, which is what is leading to our problems. The agent root directory is probably too low in the chain, as a search from there would start locating files in nearby runs I believe.

The agent working directory is a little unclear to me, the only citation I can find is "The working directory for this agent." and the path doesn't match the expected file locations.

While this has been a wonderful learning adventure, I think I'm stuck.

In summary, I've taken this about as far as I can, and someone more familiar with the project and the agent directory structure expectations should probably address this instead.

Let me know if any further information would be helpful.

**Comment by ColinDabritz** [comment]
Ok, for reference here is the workaround I'm going with.

We have a specific challenge because in the tooling included with Xamarin Studio 6 etc the IPA output is now in a folder that included a date time stamp, so it's not easy to specify the location as it changes with each build.

To work around this problem, we use a PowerShell step to resolve the path, and export the full path with no wildcards.

1. Create a release variable to hold the full path, such as `appfullpath` I used a clearly default value, such as `appfullpath variable is unresolved` so it will be clear if it did not get set.
2. Use a powershell inline script (or full script in your source) to resolve the path, and then export the value with the vso task set variable command as documented here: https://github.com/Microsoft/vsts-tasks/blob/master/docs/authoring/commands.md

    Our script looked like:
    ```powershell
    $ipa = Resolve-Path SomeProject.iOS\bin\iPhone\UITest\**\*.ipa
    $ipaPath = $ipa.Path
    echo "Full IPA path:"
    echo $ipaPath
    Write-Host ("##vso[task.setvariable variable=appfullpath;]$ipaPath")
    ```
3. In the Xamarin TestCloud step for the App File field, use the variable from step one that now has the fully qualified value, such as `$(appfullpath)`
4. You may need to fully qualify the test-cloud.exe path as well, which can be done manually in the [...] dialog.

With powershell fully resolving the wildcards as expected, and the TestCloud step only being given fully qualified paths, this seems like a clean workaround for the time being.

**Comment by madhurig** [comment]
@ColinDabritz: Thanks for reporting the issue with details and a workaround. We will work on a fix in the task.

**Comment by ColinDabritz** [comment]
Thanks much, I was hoping to be able to help further, maybe next time. It's great to have such an open and responsive Microsoft, out ream is really loving it!

**Comment by ColinDabritz** [comment]
I also realized I should add, the same issue affects the default test-cloud.exe location as well. In my scenario I can simply specify a full path easily enough, although this would mean the release would break on any upgrade of the package version. This issue may affect other step locations that are assuming a build environment instead of a release environment.

**Comment by madhurig** [comment]
The fix is merged to master but it won't be available in Team Services till around Aug 15. @ColinDabritz: Thank you for reporting the issue with great detail. I was able to repro it and verify my fix.

**Comment by ColinDabritz** [comment]
Thanks for the fix, looking forward to it! 👍

**Comment by jamesatgithub** [comment]
@madhurig is this fix available on VSTS? Or I need do any update on our macincloud build agent?

As I encountered similar problem with wildcard of the .ipa location on Test Cloud iOS task:

"No matching app files were found with search pattern: /Users/vso113024/vsts-agent/_work/5/s/Mobile/App/bin/iPhone/Release/**/*.ipa
Unhandled: tl.exit is not a function"

**Comment by madhurig** [comment]
@jamesatgithub: The fix was deployed several months back to VSTS. You should not need any changes to be made on the MacInCloud agent. Can you please open a new issue with your build log or post your log here?

**Comment by jamesatgithub** [comment]
@madhurig 
logs of two of attempts of using different wild card path copied below:

2017-02-10T01:59:15.2404750Z ##[section]Starting: Test Mobile/App.IOS/bin/iPhone/Release/**/*.ipa with Xamarin.UITest in Xamarin Test Cloud
2017-02-10T01:59:15.3112540Z ==============================================================================
2017-02-10T01:59:15.3127260Z Task         : Xamarin Test Cloud
2017-02-10T01:59:15.3141450Z Description  : Test mobile apps with Xamarin Test Cloud using Xamarin.UITest
2017-02-10T01:59:15.3155620Z Version      : 1.0.31
2017-02-10T01:59:15.3170850Z Author       : Microsoft Corporation
2017-02-10T01:59:15.3185370Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613744)
2017-02-10T01:59:15.3199850Z ==============================================================================
2017-02-10T01:59:15.8439190Z ##[error]No matching app files were found with search pattern: /Users/vso113024/vsts-agent/_work/5/s/Mobile/App.iOS/bin/iPhone/Release/**/*.ipa
2017-02-10T01:59:15.8474470Z ##[error]Unhandled: tl.exit is not a function
2017-02-10T01:59:15.9362700Z ##[section]Finishing: Test Mobile/App.IOS/bin/iPhone/Release_/**/*.ipa with Xamarin.UITest in Xamarin Test Cloud


2017-02-09T18:06:13.6001220Z ##[section]Starting: Test **/Release/**/*.ipa with Xamarin.UITest in Xamarin Test Cloud
2017-02-09T18:06:13.6699190Z ==============================================================================
2017-02-09T18:06:13.6713940Z Task         : Xamarin Test Cloud
2017-02-09T18:06:13.6728050Z Description  : Test mobile apps with Xamarin Test Cloud using Xamarin.UITest
2017-02-09T18:06:13.6741710Z Version      : 1.0.31
2017-02-09T18:06:13.6755160Z Author       : Microsoft Corporation
2017-02-09T18:06:13.6768890Z Help         : [More Information](https://go.microsoft.com/fwlink/?LinkID=613744)
2017-02-09T18:06:13.6782910Z ==============================================================================
2017-02-09T18:06:15.7669630Z ##[error]No matching app files were found with search pattern: /Users/vso113024/vsts-agent/_work/5/s/**/Release/**/*.ipa
2017-02-09T18:06:15.7707360Z ##[error]Unhandled: tl.exit is not a function
2017-02-09T18:06:15.8152020Z ##[section]Finishing: Test **/Release/**/*.ipa with Xamarin.UITest in Xamarin Test Cloud

**Comment by madhurig** [comment]
@jamesatgithub: It looks like the wild card pattern is not correct. Can you try setting the IPA path to Mobile/App.iOS/**/*.ipa or just **/*.ipa and see if the task finds the IPA. 

Thanks,
Madhuri

**Comment by jamesatgithub** [comment]
@madhurig yes **/*.ipa works. I wanted to use a path like **/$(BuildConfiguration)/**/*.ipa as we have several build configuration and need to pick up the correct ipa to send to Test Cloud, but this path didn't work. Is the path wrong or not supported for my case?

**Comment by madhurig** [comment]
@jamesatgithub: That should work too as long as the value of $(BuildConfiguration) is in the path to the .ipa. It works in my build for a sample application as expected. **/$(BuildConfiguration)/**/*.ipa finds "/Users/madhurig/coreCLRAgent/2.111.1/_work/1/s/Xamarin/Quickstarts/CreditCardValidator.iOS/CreditCardValidator.iOS/bin/iPhone/**Release**/CreditCardValidatoriOS 2017-02-20 12-28-01/**CreditCardValidatoriOS.ipa**"

Try setting a variable system.debug = true on your build definition to get a detailed build log. That will show more information on the paths searched and patterns matched.
